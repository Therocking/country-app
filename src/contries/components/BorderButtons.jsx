import { useNavigate } from 'react-router-dom';
import { useCountriesSlice } from '../../hooks/useCountriesSlice'

export const BoderButtons = ({borders}) => {
  const navigate = useNavigate();

  const { startGetCountryBorder } = useCountriesSlice();
  
  const filterCountryBorder = async(countryBorder) => {
    const country = await startGetCountryBorder( countryBorder );
	  
    const countryName = country[0].name.common;

    navigate(`/country/${countryName}`);
  }

  return (
    <>
        <button onClick={() => filterCountryBorder(borders) }>
            {borders}
        </button>
    </>
  )
}
