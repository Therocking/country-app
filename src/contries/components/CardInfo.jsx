import { useMemo } from 'react';
import { useNavigate } from 'react-router-dom'


export const CardInfo = ({name, population, region, capital, flags}) => {
  const navigate = useNavigate();
  // console.log(capital[0]);

  const onSelectCountry = () => {
   navigate(`/country/${name.common}`)
  }

  const populationString = useMemo(() => population.toLocaleString(), [population])
  
  return (
    <div onClick={ onSelectCountry } aria-label="card-info" className="card-info">
  
          <figure>
              <img src={flags.png} loading="lazy" alt={`${name.common}.png`} />
          </figure>

          <article className="info-contry">
              <h4>{name.common}</h4>

              <p>populaton: <span>{populationString}</span> </p>
              <p>region: <span>{region}</span> </p>
              <p>capital:          {
                                      capital?.map( cap => <span key={cap}> { cap } </span> )
                                   } </p>
          </article>
    </div>
  )
}
