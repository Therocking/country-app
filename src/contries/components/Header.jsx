import { Moon } from '../icons/Moon'

export const Header = () => {
    const setColorTheme = () => {
        const theme = document.body.dataset.theme

	// Arreglar esta condicion
	//if( !localStorage.getItem('theme') {
	//	document.querySelector('body').setAttribute('data-theme', 'light');
	//}
	if( theme === 'dark' ) {
            
            localStorage.setItem('theme', 'light')
            document.querySelector('body').setAttribute('data-theme', 'light');
        }else {

            localStorage.setItem('theme', 'dark')
            document.querySelector('body').setAttribute('data-theme', 'dark');        
        }
    }

  return (
    <header className="header">
        <div className="div-title">
            <h2>Where in the world?</h2>
        </div>

        <div className="container-switcher">
            <div className="container-icon">
                <Moon/>
            </div>

            <button onClick={setColorTheme} className='btn-dark-mode'>
                <span>dark mode</span>
            </button>
        </div>
    </header>
  )
}
