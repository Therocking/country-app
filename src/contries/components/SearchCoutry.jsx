import { useCountriesSlice } from '../../hooks/useCountriesSlice';
import { Search } from '../icons/Search'

export const SearchCoutry = () => {

  const {  startFilterCountries } = useCountriesSlice();

  const onFilterCountry = (e) => {
    const countryName = e.target.value.toLowerCase(); 

    startFilterCountries(countryName);
  }


  return (
    <div className="container-input">
        <div className="container-icon-search">
            <Search/>
        </div>
        
        <input 
          onChange={ onFilterCountry }
          type="text" 
          className="input" 
          placeholder="Search for a country..."
        />
    </div>
  )
}
