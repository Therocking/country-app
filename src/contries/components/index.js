

export * from './CardContry'
export * from './Header'
export * from './CountryDetailInfo'
export * from './SelectRegion'
export * from './SearchCoutry'