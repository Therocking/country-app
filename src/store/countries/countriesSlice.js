import { createSlice } from '@reduxjs/toolkit';

export const countriesSlice = createSlice({
name: 'countries',
initialState: {
    isLoading: true,
    countries: [],
    countrySelected: [],
    error: null,
},
reducers: {
    getCountries: (state, {payload}) => {
        state.isLoading = false;
        state.countries = payload;
    },
    getCoutryByRegion: (state, {payload}) => {
        state.isLoading = true;
        state.countries = payload;
    },
    getCoutrySearched: (state, {payload}) => {
        state.isLoading = true;
        state.countries = payload;
    },
    getCountrySelected: (state, {payload}) => {
        state.isLoading = true;
        state.countrySelected = payload;
    },
    endLoading: (state) => {
        state.isLoading = false;
    },
    getError: (state, {payload}) => {
        state.error = payload;
    },
},
});
export const { getCountries, endLoading,  getError, getCoutryByRegion, getCoutrySearched, getCountrySelected } = countriesSlice.actions;