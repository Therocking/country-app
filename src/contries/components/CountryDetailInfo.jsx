import { useEffect, useMemo } from 'react';
import { chageTitle, /* changeIcon */ } from '../../helpers';
import { BoderButtons } from './BorderButtons';

export const CountryDetailInfo = ({country}) => {

    const {
        name, 
        flags, 
        population, 
        region, 
        subregion,
        independent,
        capital,
        car, 
        borders
    } = country;

    useEffect(() => {
        chageTitle(name.common);
    }, [])

    const notHave = useMemo(() => <span>Not Have</span>, []);

    const populationString = useMemo(() => population.toLocaleString(), [])

    
    return (
        <div className="flex-details">
            <figure className="container-img">
                <img src={flags.png} loading="lazy" alt={`${'name.common'}img`} />
            </figure>

            <article className="details">
                <h2>{name.common}</h2>

                <div className="flex-article">
                    <p>native name:      <span>{name.official}</span> </p>
                    <p>population:       <span>{populationString}</span> </p>
                    <p>region:           <span>{region}</span> </p>
                    <p>sub region:       <span>{subregion}</span> </p>
                    <p>independent:       <span>{JSON.stringify(independent)}</span> </p>
                    <p>side cars:       <span>{car.side}</span> </p>
                    <p>capital:          {
                                            capital.map( cap => <span key={cap}> {cap} </span> )
                                         } </p>
                    
                </div>

                    <p>border contries:</p>
                <div className="borderContries">
                    {
                        (borders === undefined)
                        ? notHave
                        : borders.map( border => (
                            <BoderButtons 
                                key={border}
                                borders={border}
                            />
                        ) )
                    }
                </div>
            </article>
        </div>
  )
}
