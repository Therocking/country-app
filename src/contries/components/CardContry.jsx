import { useEffect } from 'react';
import { useCountriesSlice } from '../../hooks/useCountriesSlice'
import { CardInfo } from './CardInfo'

export const CardContry = () => {
  const {  startGetCountries, startFilterCountriesByRegion, countries, isLoading } = useCountriesSlice();


  /* Revisa el localstorage para llamar la funcion */
  useEffect(() => {
	  const region = localStorage.getItem("region");
	if( region ) {
		startFilterCountriesByRegion(region);
	}else {
    		startGetCountries();   
	}
  }, [])

  return (
    <section className="card">
      {
        (isLoading) 
        ? <h2>loading..</h2>
        : countries.map( country => (
            <CardInfo
              key={country.name.common}
              {...country}
            />
          ))
      }
    </section>
  )
}
