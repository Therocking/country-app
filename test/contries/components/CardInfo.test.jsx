import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { render, screen } from "@testing-library/react";

import { CardInfo } from "../../../src/contries/components/CardInfo"
import { countriesSlice } from "../../../src/store/countries/countriesSlice";


const store = configureStore({
  reducer: {
      auth: countriesSlice.reducer
  },
});


describe('pruebas en <CardInfo/>', () => {

  const props = {
    name: 'alemania' , 
    population: 456454,
    region: 'europe', 
    capital: 'berlin', 
    flags: {
      png: 'htttp://unabandera.com'
    }
  }

  test('debe de mostrar el componente', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>

          <CardInfo name={props.name} population={props.population} region={props.region} capital={props.capital} flags={props.flags} />
        </MemoryRouter>
      </Provider>
    );
    
    screen.debug();


    expect( props.name ).toBeTruthy()
  })

  test('debe hacer match con el snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>

          <CardInfo name={props.name} population={props.population} region={props.region} capital={props.capital} flags={props.flags} />
        </MemoryRouter>
      </Provider>
    );

    expect(container.innerHTML).toMatchSnapshot()

 })
  
})
