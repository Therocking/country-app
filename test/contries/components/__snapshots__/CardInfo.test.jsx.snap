// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <CardInfo/> debe hacer match con el snapshot 1`] = `"<div aria-label="card-info" class="card-info"><figure><img src="htttp://unabandera.com" loading="lazy" alt="alemania.png"></figure><article class="info-contry"><h4>alemania</h4><p>populaton: <span>456,454</span> </p><p>region: <span>europe</span> </p><p>capital: <span>berlin</span> </p></article></div>"`;
